### PR TITLE
Fast Typing for application rule

### PR DESCRIPTION
**Pull request by ppedrot**

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1342472948

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-solvable │  169.92   172.37  -1.42 │   473662805335    479745748019  -1.27 │   673018313699    683931003589  -1.60 │  672876   672776   0.01 │   0    0    nan │
│          coq-mathcomp-field │  222.23   224.98  -1.22 │   618903313406    626052569469  -1.14 │   943898790300    955917116663  -1.26 │  670072   662752   1.10 │   0    0    nan │
│      coq-mathcomp-character │  150.66   152.50  -1.21 │   418813808684    423980132991  -1.22 │   588885577192    598900824248  -1.67 │  718636   731128  -1.71 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1054.98  1065.92  -1.03 │  2937842975305   2968090251064  -1.02 │  4684912572790   4720451951326  -0.75 │  939900   904152   3.95 │   0    0    nan │
│                 coq-unimath │ 4357.42  4395.65  -0.87 │ 12126569574531  12231335511426  -0.86 │ 23347532694409  23435595648062  -0.38 │ 3262988  3265472  -0.08 │   0    0    nan │
│  coq-performance-tests-lite │ 1639.60  1650.28  -0.65 │  4574390699049   4603423153189  -0.63 │  7425226914971   7469764487554  -0.60 │ 2079088  2079148  -0.00 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  746.02   750.68  -0.62 │  2077592013939   2090077835614  -0.60 │  3267409796374   3288219579697  -0.63 │ 1071208  1071456  -0.02 │   0    0    nan │
│        coq-mathcomp-algebra │  141.50   142.36  -0.60 │   392982592099    396291435378  -0.83 │   513715367603    519239393704  -1.06 │  553460   553132   0.06 │   0    0    nan │
│                coq-bedrock2 │  309.98   311.78  -0.58 │   859250483666    864300257480  -0.58 │  1513740783805   1515087707940  -0.09 │  798328   798244   0.01 │   0    0    nan │
│                 coq-coqutil │   58.37    58.64  -0.46 │   162841128137    162630019750   0.13 │   214919456056    214958603469  -0.02 │  599344   598048   0.22 │   0    0    nan │
│                    coq-core │  109.67   110.05  -0.35 │   304383549911    304189557445   0.06 │   457928961225    457765751668   0.04 │  301960   301960   0.00 │   0    0    nan │
│                    coq-hott │  348.37   349.56  -0.34 │   966058806984    969008579386  -0.30 │  1406804611698   1410469710109  -0.26 │  568696   568556   0.02 │   0    0    nan │
│       coq-mathcomp-fingroup │   48.60    48.74  -0.29 │   134074894469    135266617296  -0.88 │   187373193566    190549543765  -1.67 │  480356   479600   0.16 │   0    0    nan │
│                coq-rewriter │  676.39   678.17  -0.26 │  1877198020630   1884889154792  -0.41 │  2902670897188   2919620772243  -0.58 │ 1071396  1071336   0.01 │   0    0    nan │
│                coq-coqprime │  328.34   329.15  -0.25 │   913013827748    915589456574  -0.28 │  1597363574133   1599131425732  -0.11 │  855752   855564   0.02 │   0    0    nan │
│                   coq-verdi │   98.12    98.36  -0.24 │   270527793385    270962695873  -0.16 │   381315393687    382755746568  -0.38 │  543172   542172   0.18 │   0    0    nan │
│                  coq-stdlib │  422.87   423.76  -0.21 │  1184675233032   1185624597114  -0.08 │  1492614242424   1495431279531  -0.19 │  596700   596316   0.06 │   0    0    nan │
│       coq-engine-bench-lite │  315.09   315.74  -0.21 │   889127907347    888664438166   0.05 │  1654448696578   1661432376418  -0.42 │ 3506380  3506320   0.00 │   0    0    nan │
│             coq-fiat-crypto │ 4876.52  4886.12  -0.20 │ 13544947667819  13573493513143  -0.21 │ 23943025702233  23983702924048  -0.17 │ 2217612  2234540  -0.76 │   0    0    nan │
│              coq-verdi-raft │ 1114.12  1116.23  -0.19 │  3101952707630   3108773657206  -0.22 │  4390383957504   4407555238444  -0.39 │  900560   900432   0.01 │   0    0    nan │
│                   coq-color │  486.52   487.11  -0.12 │  1359450731517   1358629472141   0.06 │  1710766880659   1712863955804  -0.12 │ 1142452  1142432   0.00 │   0    0    nan │
│                  coq-geocoq │ 1342.87  1344.23  -0.10 │  3740471250630   3744964909648  -0.12 │  5487858385985   5489348829675  -0.03 │  982348  1003604  -2.12 │   0    0    nan │
│                    coq-corn │ 1496.16  1497.48  -0.09 │  4166327070603   4170232756381  -0.09 │  6015378876911   6028540923750  -0.22 │  810000   809804   0.02 │   0    0    nan │
│             coq-lambda-rust │ 1267.83  1268.69  -0.07 │  3527336925990   3530189551964  -0.08 │  5094386382184   5101417389901  -0.14 │ 1097036  1077248   1.84 │   0    0    nan │
│               coq-perennial │ 6827.56  6824.09   0.05 │ 19022380613509  19007766134222   0.08 │ 29159988090506  29189599583153  -0.10 │ 3086704  3089744  -0.10 │   0    0    nan │
│      coq-mathcomp-ssreflect │   54.25    54.21   0.07 │   147539379891    148335321153  -0.54 │   190969542070    192109864624  -0.59 │  528744   528608   0.03 │   0    0    nan │
│            coq-fiat-parsers │  591.41   590.95   0.08 │  1654735441262   1653816301001   0.06 │  2546083636816   2548285607781  -0.09 │ 3076096  3076132  -0.00 │   0    0    nan │
│               coq-fourcolor │ 2551.38  2549.26   0.08 │  7102059165416   7105976221678  -0.06 │ 12245665400810  12249634584963  -0.03 │  731072   731048   0.00 │   0    0    nan │
│               coq-fiat-core │  108.76   108.65   0.10 │   306594867139    306846103348  -0.08 │   413787242472    413036471572   0.18 │  480372   480292   0.02 │   0    0    nan │
│            coq-math-classes │  176.54   176.36   0.10 │   492208007327    492042272589   0.03 │   652332822037    653864662370  -0.23 │  481712   481972  -0.05 │   0    0    nan │
│              coq-coquelicot │   68.60    68.44   0.23 │   186991617382    187364594909  -0.20 │   240118533835    240172711267  -0.02 │  556764   555844   0.17 │   0    0    nan │
│                 coq-bignums │   57.60    57.28   0.56 │   159560679582    159040457417   0.33 │   209059408272    209082624985  -0.01 │  465340   464896   0.10 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
Not so bad. It changes the complexity class of the algorithm from quadratic to superlinear or so, so it's a net improvement.

**Comment by JasonGross** [comment]
I imagine that if you split out coq-performance-tests-lite into per-file measurements, you'll find the 10s speedup concentrated in a single file or so, and it'll look more impressive?

**Comment by ppedrot** [comment]
@JasonGross no, even for this development the speedup is spread out. I don't know whether we have a bench precisely for typing. Note that pretyping is still quadratic so we're not seeing anything for this situation.

**Comment by JasonGross** [comment]
Ah, I see, and that's too bad that pretyping is still quadratic

**Comment by ppedrot** [comment]
This needs an assignee. @SkySkimmer ?

File: pretyping/typing.ml [hunk 1]
```diff
@@ -59,57 +59,54 @@ let assumption_of_judgment env sigma j =
...
+    | Prod (_, c1, c2) ->
+      (* Fast path *)
+      let nsubs = Esubst.subs_cons (Vars.make_substituend hj.uj_val) subs in
+      decompose sigma ((hj.uj_type, subs, c1) :: accu) nsubs c2 restjl
+    | _ ->
+      let typ = Vars.esubst Vars.lift_substituend subs typ in
+      let subs = Esubst.subs_id 0 in
+      let nsubs = Esubst.subs_cons (Vars.make_substituend hj.uj_val) subs in
+      let sigma, (c1,c2) =
+        match EConstr.kind sigma (whd_all env sigma typ) with
```
Comment (SkySkimmer) [inline pretyping/typing.ml:78]: IIUC here we may be reducing an ill-typed term, and so possibly not terminate. Is that not a concern?

Comment (ppedrot) [reply pretyping/typing.ml:78]: Hm, this is right. I think I can partially revert the code splitting the evaluation in two phases and try to be more cautious about that. Non-termination is not the only potential issue, we might get into weird crashes / anomalies.

**Comment by ppedrot** [comment]
I've reinstated the interleaving between checking the domain and deconstructing the product. This should solve @SkySkimmer 's concerns.

**Review by SkySkimmer — APPROVED** [review]
Do you want to rebench?

**Comment by ppedrot** [comment]
If you want: https://gitlab.com/coq/coq/-/jobs/1355130531. I don't expect any change, profiling I've done before showed that it never happens to have a failing typing due to arity mismatch.

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-solvable │  169.29   171.60  -1.35 │   471067067465    476864428946  -1.22 │   668194200777    680202410959  -1.77 │  679452   672744   1.00 │   0    0    nan │
│      coq-mathcomp-character │  149.91   151.59  -1.11 │   416487180559    421317883260  -1.15 │   584822566435    595888266789  -1.86 │  714672   723844  -1.27 │   0    0    nan │
│       coq-mathcomp-fingroup │   48.02    48.34  -0.66 │   132973397318    134415008101  -1.07 │   186055236391    189475909322  -1.81 │  481248   479832   0.30 │   0    0    nan │
│          coq-mathcomp-field │  221.28   222.56  -0.58 │   616129899458    620145046309  -0.65 │   940234842735    953152478655  -1.36 │  671076   670300   0.12 │   0    0    nan │
│        coq-mathcomp-algebra │  141.19   141.80  -0.43 │   391907398036    394352645030  -0.62 │   511022313578    517032374556  -1.16 │  553768   555924  -0.39 │   0    0    nan │
│              coq-coquelicot │   67.88    68.15  -0.40 │   186171258839    186487683300  -0.17 │   239302534720    239584458807  -0.12 │  557724   555768   0.35 │   0    0    nan │
│  coq-performance-tests-lite │ 1587.32  1593.19  -0.37 │  4429866384679   4444056370147  -0.32 │  7337207889235   7388785453636  -0.70 │ 2079240  2079192   0.00 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │  742.82   745.40  -0.35 │  2067245384297   2075306446583  -0.39 │  3223135158043   3244037899887  -0.64 │ 1100520  1124648  -2.15 │   0    0    nan │
│            coq-fiat-parsers │  588.25   589.84  -0.27 │  1646148948365   1647848750131  -0.10 │  2527762809766   2531906608025  -0.16 │ 3475356  3076100  12.98 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1051.01  1053.80  -0.26 │  2926245914258   2933215490107  -0.24 │  4667180596701   4705943498464  -0.82 │  909916   939348  -3.13 │   0    0    nan │
│      coq-mathcomp-ssreflect │   53.54    53.68  -0.26 │   146533221564    146896512150  -0.25 │   189997853839    191540624183  -0.81 │  525316   528540  -0.61 │   0    0    nan │
│               coq-fourcolor │ 2547.34  2553.26  -0.23 │  7098451695852   7107430612720  -0.13 │ 12239599957784  12246450101392  -0.06 │  731000   730856   0.02 │   0    0    nan │
│              coq-verdi-raft │ 1108.86  1110.91  -0.18 │  3089348025905   3093799038279  -0.14 │  4351798554753   4371944847634  -0.46 │  904352   904848  -0.05 │   0    0    nan │
│                 coq-unimath │ 4260.95  4263.12  -0.05 │ 11860650771111  11863685471816  -0.03 │ 22742027211982  22822876486310  -0.35 │ 3283452  3263600   0.61 │   0    0    nan │
│                    coq-hott │  339.86   339.90  -0.01 │   938747711396    937919084198   0.09 │  1396035586516   1400391410942  -0.31 │  569780   569784  -0.00 │   0    0    nan │
│               coq-perennial │ 6765.41  6763.51   0.03 │ 18844982842084  18838943264936   0.03 │ 28785325646104  28844403360143  -0.20 │ 3081460  3117316  -1.15 │   0    0    nan │
│             coq-lambda-rust │ 1264.88  1264.48   0.03 │  3519857507396   3519255728476   0.02 │  5066199413769   5077756649036  -0.23 │ 1097072  1096728   0.03 │   0    0    nan │
│                coq-bedrock2 │  310.21   309.97   0.08 │   858940576729    858409412729   0.06 │  1508605871887   1510079442298  -0.10 │  798376   798212   0.02 │   0    0    nan │
│             coq-fiat-crypto │ 4837.78  4832.97   0.10 │ 13434547203042  13426161086174   0.06 │ 23736616258203  23779678103611  -0.18 │ 2225492  2225360   0.01 │   0    0    nan │
│                    coq-corn │ 1476.28  1474.16   0.14 │  4108789010119   4107722475993   0.03 │  5904695634927   5919784774460  -0.25 │  813156   813008   0.02 │   0    0    nan │
│       coq-engine-bench-lite │  314.87   314.38   0.16 │   887083628541    884873332315   0.25 │  1650793125689   1660368256480  -0.58 │ 3506084  3506276  -0.01 │   0    0    nan │
│                   coq-color │  486.00   485.18   0.17 │  1355188784122   1353434957857   0.13 │  1702536193982   1705911947638  -0.20 │ 1161732  1160452   0.11 │   0    0    nan │
│                coq-rewriter │  670.57   669.33   0.19 │  1859534604168   1859481791014   0.00 │  2851621362163   2870171961713  -0.65 │ 1071340  1071700  -0.03 │   0    0    nan │
│                 coq-coqutil │   58.81    58.66   0.26 │   162592385028    162855538010  -0.16 │   214581966090    214638864980  -0.03 │  600384   598668   0.29 │   0    0    nan │
│            coq-math-classes │  173.94   173.40   0.31 │   485740480675    485526463718   0.04 │   644167960687    645789519000  -0.25 │  481112   481160  -0.01 │   0    0    nan │
│                 coq-bignums │   56.38    56.19   0.34 │   155514216797    154552454543   0.62 │   208718956320    208761989979  -0.02 │  465176   465044   0.03 │   0    0    nan │
│                  coq-stdlib │  416.44   414.84   0.39 │  1164152224140   1160796564299   0.29 │  1480032365743   1483974160316  -0.27 │  597460   596684   0.13 │   0    0    nan │
│                  coq-geocoq │ 1336.39  1330.44   0.45 │  3726130582829   3713034001390   0.35 │  5471347492331   5472315957636  -0.02 │ 1067672   992500   7.57 │   0    0    nan │
│                   coq-verdi │   98.93    98.46   0.48 │   271963800664    270617597246   0.50 │   380040175061    381657581024  -0.42 │  547652   542496   0.95 │   0    0    nan │
│                    coq-core │  106.75   106.24   0.48 │   296350286791    295798232425   0.19 │   458351030332    458254565215   0.02 │  283888   283896  -0.00 │   0    0    nan │
│                coq-coqprime │  329.46   327.52   0.59 │   916001112465    910748773852   0.58 │  1596120525533   1596137340717  -0.00 │  913944   914608  -0.07 │   0    0    nan │
│               coq-fiat-core │  108.91   107.90   0.94 │   306703641622    305186161663   0.50 │   412906684406    412157101728   0.18 │  480212   479616   0.12 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
*Tot ben.*

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Pull request merged**